extensions/amp-date-countdown/0.1/amp-date-countdown.js

```diff
@@ -0,0 +1,261 @@
+    this.endDate_ = '';
+
+    /** @private {number} */
+    this.timestampMs_ = 0;
+
+    /** @private {number} */
+    this.timestampSeconds_ = 0;
+
+    /** @private {number} */
+    this.offsetSeconds_ = 0;
+
+    /** @private {string} */
+    this.biggestUnit_ = '';
+
+    /** @const {!../../../src/service/template-impl.Templates} */
+    this.templates_ = Services.templatesFor(this.win);
+  }
+
+  /** @override */
+  buildCallback() {
+    const self = this;
+    //Note: One of datetime, timestamp-ms, timestamp-seconds is required.
+    this.endDate_ = this.element.getAttribute('end-date');
+    this.timestampMs_ = Number(this.element.getAttribute('timestamp-ms'));
+    this.timestampSeconds_
+      = Number(this.element.getAttribute('timestamp-seconds'));
+    this.offsetSeconds_ = Number(this.element.getAttribute('offset-seconds'));
+
+    const locale = this.element.getAttribute('locale') || 'en';
+    const whenEnded = this.element.getAttribute('when-ended') || 'stop';
+    this.biggestUnit_ = (this.element.getAttribute('biggest-unit')) || 'days';
+
+    let epoch = this.getEpoch_(); //1519776000000
+    if (this.offsetSeconds_) {
+      epoch += this.offsetSeconds_ * 1000;
+    }
+    let differentBetween = new Date(epoch) - new Date();
+    let data;
+    const localeWordList = this.getLocaleWord_(locale);
+    const delay = 1000;
+
+    const countDownTicker = setInterval(function() {
+      //console.log('differentBetween', differentBetween);
+      data = self.getYDHMSFromMs_(differentBetween);
+      if (whenEnded === 'stop' && differentBetween < 1000) {
+        Services.actionServiceForDoc(self.element)
+            .trigger(self.element, 'timeout', null, ActionTrust.HIGH);
+        clearInterval(countDownTicker);
+      }
+      differentBetween -= delay;
+      self.renderItems_(Object.assign(data, localeWordList));
+    }, delay);
+
+    this.registerAction('timeout', this.timeout_.bind(this));
+
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /**
+   * @param {!Object} items
+   * @return {!Promise}
+   * @private
+   */
+  renderItems_(items) {
+    return this.templates_.findAndRenderTemplate(this.element, items)
+        .then(this.boundRendered_);
+  }
+
+  /**
+   * @return {number}
+   * @private
+   */
+  getEpoch_() {
+    let epoch;
+
+    if (this.endDate_) {
+      epoch = Date.parse(this.endDate_);
+    } else if (this.timestampMs_) {
+      epoch = this.timestampMs_;
+    } else if (this.timestampSeconds_) {
+      epoch = this.timestampSeconds_ * 1000;
+    }
+
+    if (isNaN(epoch)) {
+      user().error(TAG,
+          'One of end-date, timestamp-ms, timestamp-seconds is required');
+    }
+
+    //console.log('epoch', epoch);
+    return epoch;
+  }
+
+  /**
+   * Toggles the timeout_ state when timer hit 0.
+   * @param {?../../../src/service/action-impl.ActionInvocation=} opt_invocation
+   * @private
+   */
+  timeout_() {
+
```

Comment:
what should I write in this empty function ?

Reply:
I'm not sure why you register a `timeout` AMP action. I don't see it in the spec. Can you help me understand what you want the `timeout` AMP action to accomplish? Would it be used like this?

```html
<button on="tap:adc.timeout">End the countdown</button>
<amp-date-countdown id="adc"></amp-date-countdown>
```

I see you trigger a `timeout` AMP event above. Did you intend for there to be both a `timeout` AMP action and a `timeout` AMP event? There can be just an event without a matching action.

Reply:
this `timeout` event will be triggered when 0 second is hit, user might want to hide / show elements or any other actions.

Reply:
Ok I think I understand. This function and the call to `registerAction` is unnecessary to emit the timeout event. The call to ActionService's `trigger` on line 79 is all that's needed.